batcher/aligned-task-sender/src/main.rs

```diff
@@ -0,0 +1,279 @@
+                ));
+                let public_input_path =
+                    base_dir.join(format!("{}/ineq_{}_groth16.pub", GROTH_16_PROOF_DIR, count));
+                let vk_path =
+                    base_dir.join(format!("{}/ineq_{}_groth16.vk", GROTH_16_PROOF_DIR, count));
+
+                let proof = std::fs::read(&proof_path)
+                    .map_err(|e| SubmitError::IoError(proof_path, e))
+                    .unwrap();
+                let public_input = std::fs::read(&public_input_path)
+                    .map_err(|e| SubmitError::IoError(public_input_path, e))
+                    .unwrap();
+                let vk = std::fs::read(&vk_path)
+                    .map_err(|e| SubmitError::IoError(vk_path, e))
+                    .unwrap();
+                VerificationData {
+                    proving_system: ProvingSystemId::Groth16Bn254,
+                    proof,
+                    pub_input: Some(public_input),
+                    verification_key: Some(vk),
+                    vm_program_code: None,
+                    proof_generator_addr,
+                }
+            })
+        })
+        .collect();
+
+    let cached_verification_data: Vec<VerificationData> =
+        threads.into_iter().map(|t| t.join().unwrap()).collect();
+
+    // We operate over a local network each thread sources its nonce by incrementing the global network nonce.
+    // When multiple senders are spawned we just increment the atomic to grab the nonce.
+    let network = args.network.into();
+    let latest_nonce = get_next_nonce(&args.eth_rpc_url, wallet.address(), network)
+        .await?
+        .as_u64();
+    let global_nonce = Arc::new(AtomicU64::new(latest_nonce));
+    let threads = (0..args.num_senders)
+        .map(|sender_id| {
+            let batcher_url = args.batcher_url.clone();
+            let nonce = global_nonce.clone();
+            let handle = Handle::current();
+            let cached_verification_data: Vec<VerificationData> = cached_verification_data.clone();
+            let wallet = wallet.clone();
+            thread::spawn(move || {
+                info!("Task Sender Started {}", sender_id);
+                loop {
+                    let max_fees = vec![max_fee; args.burst_size];
+                    //TODO(pat): not sure how slow this is... open to faster alternatives
+                    let verification_data: Vec<_> = cached_verification_data
+                        .choose_multiple(&mut thread_rng(), burst_size)
+                        .cloned()
+                        .collect();
+                    info!(
+                        "Sending {:?} Proofs to {:?} Aligned Batcher",
+                        burst_size, network
+                    );
+                    let nonce = nonce.fetch_add(burst_size as u64, Ordering::Relaxed);
+                    let batcher_url = batcher_url.clone();
+                    let wallet = wallet.clone();
```

Comment:
I wonder if each thread could use a unique wallet. Using the same one might force serialization in the batcher if it represents the same operator.

Reply:
For this we would need to programmatically create and fund accounts on anvil for each wallet. This is possible using via two methods:
- The task sender initiates anvil with `num_senders` accounts pre-funded and extract the private keys and addresses.
- Each thread generates an address and private key pair then calls `anvil_setBalance` https://book.getfoundry.sh/reference/anvil/ .